src/main/java/com/example/gymApp/service/TrainerService.java

```diff
@@ -1,131 +1,167 @@
+  @Transactional
+  public Trainer updateTrainer(Trainer trainer, String name, String lastName, String username,
+      String password, Boolean activeStatus, String specialization) {
 
-      Trainer trainer = getTrainerById(id);
+    Optional<TrainingType> trainingTypeOptional = trainingTypeRepository.findByName(specialization);
+    if (trainingTypeOptional.isEmpty()) {
+      throw new IllegalArgumentException("Specialization not found in the database");
+    }
 
-      trainer.setFirstName(firstName);
-      trainer.setLastName(lastName);
-      trainer.setSpecialization(specialization);
+    TrainingType trainingType = trainingTypeOptional.get();
 
-      validateTrainer(trainer);
+    trainer.getUser().setFirstName(name);
+    trainer.getUser().setLastName(lastName);
+    trainer.getUser().setUsername(username);
+    trainer.getUser().setPassword(password);
+    trainer.getUser().setActive(activeStatus);
+    trainer.setSpecialization(trainingType);
 
-      // Сохраняем изменения
-      if (!trainerDAO.updateTrainer(trainer)) {
-        throw new NoSuchElementException("Trainer with ID " + id + " could not be updated.");
-      }
-      log.info("Trainer {} has been updated successfully.", id);
-      return true;
+    trainerRepository.save(trainer);
 
-    } catch (IllegalArgumentException e) {
-      log.error("Error updating trainer: {}", e.getMessage());
-      return false;
-    }
+    return trainer;
   }
 
-  public boolean deleteTrainer(Long id) {
-    try {
-      if (!trainerDAO.deleteTrainer(id)) {
-        throw new NoSuchElementException("Trainer with ID " + id + " not found.");
-      }
-      log.info("Trainer {} has been deleted successfully.", id);
-      return true;
-    } catch (NoSuchElementException e) {
-      log.error("Error deleting trainer: {}", e.getMessage());
-      return false;
+  public TrainingType checkSpecializationCorrectness(String specialization) {
+    Optional<TrainingType> trainingType = trainingTypeRepository.findByName(specialization);
+    if (trainingType.isEmpty()) {
+      throw new IllegalArgumentException("Specialization not found in the database");
     }
+    return trainingType.get(); //get clear TrainingType object;
   }
 
-  public Trainer getTrainerById(Long id) {
-    Trainer trainer = trainerDAO.getTrainerById(id);
-    if (trainer == null) {
-      throw new NoSuchElementException("Trainer with ID " + id + " not found.");
+
+  public Trainer createTrainer(String firstName, String lastName, String username, String password,
+      String specialization) {
+    if (userRepository.findByUsername(username).isPresent()) {
+      throw new IllegalArgumentException("Username of trainer already exists");
     }
+
+    TrainingType trainingType = checkSpecializationCorrectness(specialization);
+
+    User user = new User();
+    user.setFirstName(firstName);
+    user.setLastName(lastName);
+    user.setUsername(username);
+    user.setPassword(password);
+    user.setActive(true);
```

Comment:
You can put this in a separate method for creating a user and not duplicate the code.